src/main/java/org/terasology/Lakes/LakeProvider.java

```diff
@@ -61,36 +61,54 @@ public void process(GeneratingRegion region) {
         Region3i worldRegion = lakes.getWorldRegion();
         BiomeFacet biomeFacet = region.getRegionFacet(BiomeFacet.class);
 
-        for (Vector3i pos : worldRegion) {
-            float sHeight = surfaceHeightFacet.getWorld(pos.x(), pos.z());
-            float probability;
-            probability = computeProbability(pos, biomeFacet);
-            if (pos.y() < sHeight - 20 && probability > 0.999999) {
-                lakes.add(new Lake(pos, 10 + Math.round(20 * Math.abs(noise.noise(pos.x(), pos.z())))));
-            } else if (pos.y() == Math.round(sHeight) && probability > 0.9999 && checkGradient(pos,
-                    surfaceHeightFacet)) {
-                Lake temp = new Lake(pos, 10 + Math.round(20 * Math.abs((noise.noise(pos.x(), pos.z())))));
-                if (checkCorners(temp.getBoundingBox(), surfaceHeightFacet)) {
-                    int minHeight = getMinimumHeight(temp.getBoundingBox(), surfaceHeightFacet);
-                    if (minHeight < pos.y()) {
-                        temp.setWaterHeight(minHeight);
+        Vector3i min = worldRegion.min();
+        int step = 3;
+        Vector3i start = new Vector3i(
+                (min.x() + (step - Math.floorMod(min.x(), step))),
+                min.y() + (step - Math.floorMod(min.y(), step)),
+                min.z() + (step - Math.floorMod(min.z(), step))
+        );
+
+        for (int wy = start.y(); wy < worldRegion.maxY(); wy += step) {
+            for (int wx = start.x(); wx < worldRegion.maxX(); wx += step) {
+                for (int wz = start.z(); wz < worldRegion.maxZ(); wz += step) {
+                    Vector3i pos = new Vector3i(wx, wy, wz);
```

Comment:
So we just talked about this live on Discord while I was still testing this, if keeping this outside the loop caused you grief and realizing it had to be here for it to work you might want to leave a comment about it to salvage some value from that headache 😁